compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -3995,11 +3994,11 @@ public void visit(BLangObjectConstructorExpression objectCtorExpression, Analyze
         BObjectType objectType;
         if (objectCtorExpression.referenceType == null && objectCtorExpression.expectedType != null) {
             objectType = (BObjectType) objectCtorExpression.classNode.getBType();
-            if (Types.getReferredType(objectCtorExpression.expectedType).tag == TypeTags.OBJECT) {
-                BObjectType expObjType = (BObjectType) types
-                        .getReferredType(objectCtorExpression.expectedType);
+            BType effectiveType = Types.getEffectiveType(Types.getReferredType(objectCtorExpression.expectedType));
```

Comment:
Can the effective type be a reference type now since `getEffectiveType()` unwraps the intersection only now?

Reply:
Yes. is there any issue with that?

Reply:
Yes because in the next line we check for ObjectTag. For reference types that will evaluate to false

Reply:
We get the referred type(using `Types.getReferredType`) and unwrap the intersection type in L3997. Can I get an example?

Reply:
I was thinking something like
```ballerina
type Foo object {
    public string s;
};

type Bar Foo & readonly;

Bar obj = object { public string s = ""};
```
But it seems, Bar's effective type comes as an Object type instead of a type reference type due to https://github.com/ballerina-platform/ballerina-lang/issues/37097. To be on the safe side, better get the referred type anyways and check for the ObjectTag.

Reply:
Resolving this and should be handled depending on the fix for https://github.com/ballerina-platform/ballerina-lang/issues/37097